File: geometry-bughunt/findings/geos-union-drops-polygon/Repro.java
// Same two triangles through JTS (public API only).
// Build/run: javac -cp jts-core.jar Repro.java && java -cp jts-core.jar:. Repro
import org.locationtech.jts.JTSVersion;
import org.locationtech.jts.geom.Geometry;
import org.locationtech.jts.io.WKTReader;
import org.locationtech.jts.operation.overlayng.OverlayNG;
import org.locationtech.jts.operation.overlayng.OverlayNGRobust;

public class Repro {
  public static void main(String[] args) throws Exception {
    WKTReader r = new WKTReader();
    Geometry a = r.read("POLYGON ((1 1, -1e-20 0, 1 0, 1 1))");
    Geometry b = r.read("POLYGON ((0 0, 1 1, 0 1, 0 0))");
    System.out.println("JTS " + JTSVersion.CURRENT_VERSION);
    System.out.println("isValid(A)=" + a.isValid() + " isValid(B)=" + b.isValid()
        + " area(A)=" + a.getArea() + " area(B)=" + b.getArea());
    String[] names = {"union", "intersection", "B - A", "A - B", "symdifference"};
    Geometry[][] args2 = {{a, b}, {a, b}, {b, a}, {a, b}, {a, b}};
    int[] ops = {OverlayNG.UNION, OverlayNG.INTERSECTION, OverlayNG.DIFFERENCE,
                 OverlayNG.DIFFERENCE, OverlayNG.SYMDIFFERENCE};
    String[] expected = {"~1.0", "~5e-21", "~0.5", "~0.5", "~1.0"};
    for (int i = 0; i < ops.length; i++) {
      Geometry g = OverlayNGRobust.overlay(args2[i][0], args2[i][1], ops[i]);
      System.out.printf("OverlayNGRobust %-13s area = %-22s (expected %s)  %s%n",
          names[i], g.getArea(), expected[i], g);
    }
  }
}

File: geometry-bughunt/findings/geos-union-drops-polygon/repro.py
"""Shapely (GEOS) reproducer: overlay of two triangles loses most of B.

Run: python3 repro.py
Uses only Shapely's public API.
"""
import shapely
from shapely import wkt

A = wkt.loads("POLYGON ((1 1, -1e-20 0, 1 0, 1 1))")   # area 0.5 (+5e-21)
B = wkt.loads("POLYGON ((0 0, 1 1, 0 1, 0 0))")         # area 0.5

print("shapely", shapely.__version__, "GEOS", shapely.geos_version_string)
print("valid:", A.is_valid, B.is_valid, " areas:", A.area, B.area)
print("predicates: intersects", A.intersects(B), "overlaps", A.overlaps(B),
      "covers(A,B)", A.covers(B))
for name, g, exp in [
    ("union", A.union(B), "~1.0"),
    ("intersection", A.intersection(B), "~5e-21"),
    ("B - A", B.difference(A), "~0.5"),
    ("A - B", A.difference(B), "~0.5"),
    ("symdifference", A.symmetric_difference(B), "~1.0"),
]:
    print(f"{name:14s} area = {g.area!r:24s} (expected {exp})  {g.wkt}")
